woocommerce/woocommerce-ios
# [UL&S] Match store address with wordpress.com account after login is completed and present error if necessary

Closes #3482 

This is a PR against the 5.8 release branch, to fix an omission picked up during testing:

pbMoDN-1KW#comment-3684-p2
pbMoDN-1KW#comment-3687-p2

In action:

https://user-images.githubusercontent.com/2722505/104858404-cb397a80-58ec-11eb-9d8c-d3c1ad65c3bb.mov


Screenshots:

| Light Appearance | Dark Appearance |
| ---- | ---- |
| <img src="https://user-images.githubusercontent.com/2722505/104858339-4b131500-58ec-11eb-8327-7c2768886491.png" width="350"/> | <img src="https://user-images.githubusercontent.com/2722505/104858338-4b131500-58ec-11eb-95da-44ef3314f210.png" width="350"/> |
| <img src="https://user-images.githubusercontent.com/2722505/104858336-4a7a7e80-58ec-11eb-9f72-9c9fd8f1ffe6.png" width="350"/> | <img src="https://user-images.githubusercontent.com/2722505/104858335-4a7a7e80-58ec-11eb-921e-6cefe110e304.png" width="350"/> |
| <img src="https://user-images.githubusercontent.com/2722505/104858390-ae04ac00-58ec-11eb-8aed-e7b24b03bc53.png" width="350"/> | <img src="https://user-images.githubusercontent.com/2722505/104858389-ae04ac00-58ec-11eb-904e-daf9a79c0edf.png" width="350"/> |



## Rationale

When logging with Store Address, the flow is:
1. enter the store address
2. enter the email associated to the wp.com account
3. enter password
4. pick store

We missed checking if the email/password entered in steps 2 and 3 and the store address entered in step 1 correspond to the same account.

There is a serious limitation to do so. At the end of the login process, WooCommerce is notified by WordPressAuthenticator of the userId corresponding to the email/password entered in steps 2+3. However, there is no way for neither WPAuthenticator, nor WooCommerce, to know at that point the userID that corresponds to the store address entered in step 1.

## Changes
* After the login process is completed, and before presenting the store picker, we will check if the store address entered in step 1 is one of the stores associated to the wordpress.com account entered in steps 2+3. To do so, we use `ULAccountMatcher`
* Add a new error screen specific to this use case: `ULAccountMismatchViewController`. Notice there is an issue with the spacing of the "Wrong Account?" label and "LogOut" button. I can not figure out for the life of me why how to solve that gap. Any help or suggestions are more than welcome.
* Add a new view model specific for this case.
* Add tests
* Add a new configuration to StorePickerViewController to support displaying only the list of sites. StorePickerViewController had already three responsibilities at this point, and I am adding a fourth now, but I am afraid to break it down this close to a release. We should probably think of ways to simplifying it, by breaking it down into its four responsibilities.

## How to test
Prepare two different accounts with at least one store each. Let's call them Account A and Account B

First testing flow:
1. In the login prologue, tap Login With Store Address. 
2. Enter the address of a store associated to account A
3. In the next step, enter the email and password associated to account B
4. Notice if there is an error screen. 
5. In the error screen, tap the Log Out button.
6. The app should navigate to the login prologue again

Second testing flow:
1. In the login prologue, tap Login With Store Address. 
2. Enter the address of a store associated to account A
3. In the next step, enter the email and password associated to account B
4. Notice if there is an error screen. 
5. In the error screen, tap See Connected Stores.
6. The app should navigate to a list of stores associated with account B
7. Tap back, tap Log Out

Third testing flow:
1. In the login prologue, tap Login With Store Address. 
2. Enter the address of a store associated to account A
3. In the next step, enter the email and password associated to account A
4. The app should navigate to the store picker
5. Select a store, everything should work as expected.

cc: @designsimply 

Update release notes:
